File: packages/python/examples/pytest/file_upload_test.py
from os import getenv
from tempfile import NamedTemporaryFile

import pytest
from pytest import fixture, mark

from alumnium import Provider

driver_type = getenv("ALUMNIUM_DRIVER", "selenium")


@fixture
def file():
    with NamedTemporaryFile(mode="w+", suffix=".txt") as file:
        yield file


@fixture
def file2():
    with NamedTemporaryFile(mode="w+", suffix=".txt") as file:
        yield file


@mark.xfail("appium" in driver_type, reason="File upload is not implemented in Appium yet")
def test_file_upload(al, file, navigate):
    navigate("https://the-internet.herokuapp.com/upload")
    al.do(f"upload '{file.name}'")
    al.do("click on 'Upload' button")
    assert al.get("heading") == "File Uploaded!"


@mark.xfail("appium" in driver_type, reason="File upload is not implemented in Appium yet")
def test_multiple_file_upload(al, file, file2, navigate):
    if al.model.provider == Provider.AWS_META:
        pytest.xfail("Prefers to click on the upload button manually")

    navigate("multiple_file_upload.html")
    al.do(f"upload files '{file.name}', '{file2.name}'")
    al.do("click 'Upload Files' button")
    assert "✓ Upload Successful!" in al.get("success message")
    assert al.get("uploaded files names") == [
        file.name.split("/")[-1],
        file2.name.split("/")[-1],
    ]


@mark.xfail("appium" in driver_type, reason="File upload is not implemented in Appium yet")
@mark.xfail(driver_type == "selenium", reason="Hidden file upload inputs are not supported in Selenium")
def test_hidden_file_upload(al, file, navigate):
    if al.model.provider == Provider.AWS_META:
        pytest.xfail("Prefers to click on the upload button manually")

    navigate("hidden_file_upload.html")
    al.do(f"upload '{file.name}' to 'Choose Files' button")
    al.do("click 'Upload Files' button")
    assert al.get("success message") == "Files Uploaded Successfully!"

File: packages/typescript/tests/system/fileUpload.test.ts
import { randomUUID } from "node:crypto";
import fs from "node:fs/promises";
import { tmpdir } from "node:os";
import { afterEach, beforeEach, describe } from "vitest";
import { safePathJoin } from "../../src/utils/fs.ts";
import { baseIt } from "./helpers.ts";

describe("File Upload", () => {
  let testFile1: string;
  let testFile2: string;

  beforeEach(() => {
    testFile1 = safePathJoin(tmpdir(), `${randomUUID()}.txt`);
    testFile2 = safePathJoin(tmpdir(), `${randomUUID()}.txt`);
    return Promise.all([
      fs.writeFile(testFile1, "Test content 1"),
      fs.writeFile(testFile2, "Test content 2"),
    ]);
  });

  afterEach(() =>
    Promise.all([
      fs.rm(testFile1, { force: true }),
      fs.rm(testFile2, { force: true }),
    ]),
  );

  const it = baseIt.override("setup", async ({ setup, skip }) => {
    return async (options) => {
      const result = await setup(options);
      const { isMobile } = result;

      // File upload is not implemented in Appium yet
      if (isMobile) skip("File upload is not implemented in Appium yet");

      return result;
    };
  });

  it("should upload a single file", async ({ expect, setup }) => {
    const { al, $ } = await setup();

    await $.navigate("https://the-internet.herokuapp.com/upload");
    await al.do(`upload '${testFile1}'`);
    await al.do("click on 'Upload' button");

    const heading = await al.get("heading");
    expect(heading).toBe("File Uploaded!");
  });

  it("should upload multiple files", async ({ expect, setup, skip }) => {
    const { al, $, model } = await setup();

    if (model.provider === "aws_meta")
      skip("Prefers to click on the upload button manually");

    await $.navigate("multiple_file_upload.html");
    await al.do(`upload files '${testFile1}', '${testFile2}'`);
    await al.do("click 'Upload Files' button");

    const message = await al.get("success message");
    expect(message).toContain("✓ Upload Successful!");

    const uploadedFiles = await al.get("uploaded files names");
    expect(uploadedFiles).toEqual([
      testFile1.split("/").pop(),
      testFile2.split("/").pop(),
    ]);
  });

  it("should upload a hidden file", async ({ expect, setup, skip }) => {
    const { al, $, driverId: driverKind, model } = await setup();

    if (driverKind === "selenium")
      skip("Hidden file upload inputs are not supported in Selenium");

    if (model.provider === "aws_meta")
      skip("Prefers to click on the upload button manually");

    await $.navigate("hidden_file_upload.html");
    await al.do(`upload '${testFile1}' to 'Choose Files' button`);
    await al.do("click 'Upload Files' button");

    const message = await al.get("success message");
    expect(message).toBe("Files Uploaded Successfully!");
  });
});
